### PR TITLE
Support more ways to load kernel modules early

### PR DESCRIPTION
**Pull request by Foxboron**
This path implements two new ways of loading kernel modules early. Sometimes building monolithic kernels are not really feasible and people use their kernel and kernel modules included int he initrd image. This brings problems like how to supply the desired driver into `/lib/modules/*.ko` as you would need to have an explicit `-file` parameter for each module.

This introduces two new ways to get around this:
* Through configuration files in `/etc/modules-load.d`
* Through `modules_load` kernel parameter

These options are popularized through the systemd project and documented here;

https://www.freedesktop.org/software/systemd/man/kernel-command-line.html#modules_load=
https://www.freedesktop.org/software/systemd/man/modules-load.d.html

I'm also wondering if we'd be interested in supporting a new u-root argument `-module` which would create a `/etc/modules-load.d/default.conf` file in the initrd, but this can come later.

### CODE REVIEW

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:242]: "form" is a typo

Comment (abrender) [inline pkg/libinit/root_linux.go:242]: "in order:" instead of "in order;"

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:303]: Does it make sense to use strings.TrimSpace to remove leading and trailing whitespace?

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:304]: do you also want to check for empty lines?

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:310]: does it make sense to log an error if err != io.EOF?

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:326]: The log line prints "installModulesFromModulesLoad" but the function name is "InstallModulesFromModulesLoad" (capital "I" at the start)

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:332]: %q instead of %s might help here... in an edge case where the module name is empty it might not be obvious to a user reading the log line that the module name is empty

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:337]: should f.Close() be "defer f.Close()" on line 329, right after check of `err` after os.Open?

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:338]: I think you can inline this:
if err := kmodule.Probe(...); err != nil {
 // error here
}

also: the function name in the log message on line 339 is missing the initial uppercase "i"

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:347]: "pattern" isn't used in this function... is this argument needed?

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:355]: %q instead of %s here please?

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:353]: do you have to worry about whitespace? What if someone passes:
--modules_load="module_a, module_b , module_c" ?

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:360]: I think you can inline the error checking: `if err := kmodule.Probe(...); err != nil { ...`

**Comment by abrender** [comment]
Thank you for this contribution!  Can you please also add test coverage for the code? Thanks!

File: pkg/libinit/root_linux.go
Comment (Foxboron) [reply pkg/libinit/root_linux.go:353]: I would just call it invalid and move on. But we can strip both sides of the string?

Comment (abrender) [reply pkg/libinit/root_linux.go:353]: I think that calling strings.TrimSpace(moduleName) is so easy that we should just do it and help ensure a good experience for users.

**Comment by rminnich** [comment]
I have always wondered if your module loading setup could follow the standard path:
/lib/modules/kernel-verison/modules.dep for information, and that path for modules.
I don't see the point of /etc/modules.d, as part of the goal of u-root was to avoid bringing in dependencies on /etc, and it is actually redundant with what is in /lib/modules/kernel-version. 

wdyt, am I missing something?

**Comment by Foxboron** [comment]
The question currently is; if you have `/lib/modules/$kernel-$ver` and you need to load modules at loadtime. What options do you have?

Currently I can just see ``/lib/modules/*.ko` but no distro package does this, so you'd have to place them there manually. It's more convenient to just include a module directory and have the module resolution as implemented in `kmod` do it's job, right? Then you would only really need some way to tell u-root "load these modules up at statup" and `kmod` does the rest.

We can move away from `/etc/modules-load.d` and use `/lib/modules-load.d` instead.

**Comment by rminnich** [comment]
Quick question: I have this:
ls /lib/modules
5.15.0-41-generic  5.15.0-43-generic  5.15.0-46-generic

do you not have that? Maybe I'm about to learn something :-)

**Comment by rminnich** [comment]
I may be misunderstanding your comment but anyway:
on boot, 
mount proc
read /proc/version to get a version (needs a small bit of processing)
then look in /lib/modules/version

could that work?

**Comment by Foxboron** [comment]
>do you not have that? Maybe I'm about to learn something :-)

I do, but that doesn't load any modules.

>read /proc/version to get a version (needs a small bit of processing) then look in /lib/modules/version
> could that work?

`kmod` already implements this and we use that to find the modules in this implementation. The question is how you pass modulenames during boot.

**Comment by Foxboron** [comment]
Moved the module directory from `/etc/modules-load.d` to `/lib/modules-load.d` :)

File: pkg/libinit/root_linux.go
Comment (rminnich) [inline pkg/libinit/root_linux.go:353]: Right here is where you can use the new cmdline functions to create fake command lines and parse them.

Comment (rminnich) [reply pkg/libinit/root_linux.go:353]: OK, what I'd suggest is that these functions change in a few ways. We might want a struct containing needed information, and one item in that struct could be a *CmdLine created from the cmdline package. That CmdLine can then be mocked up for testing or derived from the real cmdline. This should let us right much better tests.

Comment (Foxboron) [reply pkg/libinit/root_linux.go:353]: I'll take a few stabs at this tomorrow. I'm not quite happy with this code as-is so I won't mind do a bit of reworking :)

**Comment by Foxboron** [comment]
Okay, so I have implemented test coverage for the new functionality where we load modules by reading `/proc/cmdline` and files from `/lib/modules-load.d`. This now uses a `InitialModulesLoader` struct which allows us to mock up the kmod probing as well and the `Cmdline` struct. 

I renamed the original `/lib/modules/*.ko` from `InstallModules` to `InstallModulesFromDir` and didn't integate it with `kmodule.Probe` as it is never really utilizing the probing infrastructure and do files instead.

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:361]: (https://pkg.go.dev/strings#TrimSpace) hmm so we need this because `ReadString returns err != nil if and only if the returned data does not end in delim` and the docs for TrimSpace state  `For simple uses, a Scanner may be more convenient.`

Perhaps it'll be clearer to restructure this code block as:

```
scanner := bufio.NewScanner(f)
for scanner.Scan() { // By default this uses newline delimiters
  // Do something with scanner.Text()
}
if err := scanner.Err(); err != nil {
  // Do something with err
}
```

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:379]: should this `fmt.Println` line be removed? Maybe it was left over from debugging?

File: pkg/libinit/root_linux_test.go
Comment (abrender) [inline pkg/libinit/root_linux_test.go:30]: (here and below) if you use `cmp.Diff` (https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff) you get a nice diff, instead of 2 long lists that you have to manually diff

```
if diff := cmp.Diff(expectedModules, loadedModules); diff != "" {
  t.Fatalf("unexpected difference of loaded modules (-want, +got): %v", diff)
}
```

File: pkg/libinit/root_linux_test.go
Comment (abrender) [inline pkg/libinit/root_linux_test.go:58]: I think you can replace this whole block with `T.TempDir` (https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff)

File: pkg/libinit/root_linux_test.go
Comment (abrender) [inline pkg/libinit/root_linux_test.go:67]: `WriteFile` returns an error, probably a good idea to check it

File: pkg/libinit/root_linux.go
Comment (rminnich) [reply pkg/libinit/root_linux.go]: how goes? I'd like to get this in.

File: pkg/libinit/root_linux.go
Comment (Foxboron) [reply pkg/libinit/root_linux.go:238]: We need some way to specify them, possibly outside of the library itself. So I'd rather keep it exported else we need to have a setter function as well.

**Comment by Foxboron** [comment]
I've addressed the feedback from @abrender. Thanks for the review :)

File: pkg/libinit/root_linux.go
Comment (abrender) [inline pkg/libinit/root_linux.go:354]: I think the `else` and `else if` are not needed because all the previous conditions have `continue`.  So you can either do
```
if i == "" {
  continue
}
if strings.HasPrefix(i, "#") {
  continue
}
modules = append(modules, i)
```

or just combine the `if`s:
```
if i = "" || strings.HasPrefix(i, "#") {
  continue
}
modules = append(modules, i)
```

I prefer the latter but don't have a strong preference.

File: pkg/libinit/root_linux_test.go
Comment (abrender) [inline pkg/libinit/root_linux_test.go:65]: If this t.Fatal triggers there's no indication to the user which of the multiple tests has failed.  You can fix this with "Table-driven tests using subtests" (https://go.dev/blog/subtests):

```
for _, file := range files {
  t.Run(files.Name, func(t *testing.T) {
    // Rest of existing code here
  })
}
```

File: pkg/libinit/root_linux_test.go
Comment (abrender) [inline pkg/libinit/root_linux_test.go:59]: no need to manually remove the dir:  https://pkg.go.dev/testing#B.TempDir says `The directory is automatically removed by Cleanup when the test and all its subtests complete`

**Comment by Foxboron** [comment]
Fixed up the last comments.

Should we squash the commits before merging?

**Comment by abrender** [comment]
> Fixed up the last comments.

Thanks! I don't see them pushed yet (probably related to your question below)

> 
> Should we squash the commits before merging?

I'll defer to @rminnich.

**Comment by Foxboron** [comment]
Sorry :) Amended the commit and everything should be there now

**Comment by abrender** [comment]
Awesome, thanks for this contribution! Looks good from me on the code side... I'll leave it to @rminnich for further review/merge.

File: pkg/libinit/root_linux.go
Comment (rminnich) [inline pkg/libinit/root_linux.go:237]: one little nit: it's frequently best to define an interface, and the create types like this struct that implement the interface. Why? b/v freebsd has modules, just not modules like this. Also, mocking for testing is easier. You can see this in https://github.com/u-root/u-root/blob/main/pkg/efivarfs/vars_test.go#L16, where I was able to put together things that allowed tests to be easy.

**Comment by rminnich** [comment]
So I know you need this, and testing the functionality is tricky, since it runs as root and adds modules to the kernel. All that said, you might want to think hard about upping the testing. Approving since it's critical infra.

**Review by rminnich — APPROVED** [review]

**Comment by rminnich** [comment]
squashing the commits first would be a good idea, I just noticed titles like 'fix up' :-)

**Review by rminnich — CHANGES_REQUESTED** [review]
makes sense to squash them, then we should be ready to go.

**Comment by Foxboron** [comment]
So I understand the need for better test coverage. But it's admittedly a bit hard to do full integration testing since we don't really deal with the `kmod` portion of the code. I'm a tiny bit unsure what more test coverage would get us as we are in practise only really implementing two different parsers.

The commits have been squased :) I'm also around OSFC

**Review by rminnich — APPROVED** [review]

**Pull request merged**
